### PR TITLE
[installer-preview] add werft flags for triggering self-hosted preview

### LINKED ISSUE #7824: Automate builds of self-hosted Gitpod for preview environments
For validating features like self-hosted admin and auth integrations, it would be useful if developers could build and deploy dedicated preview environments with a self-hosted Gitpod installation, the same way we currently do for SaaS.

Maybe this could be controlled with a label or /command in the PR description, to override the default preview build and deployment.

**Typical use-case**  
- Create a PR with new self-hosted admin dashboard feature to configure auth integration.
- This triggers werft jobs to build and deploy a self-hosted instance in a preview environment.
- On commit, re-run the build and redeploy self-hosted to the same preview environment.

cc: @svenefftinge

### PR DESCRIPTION
**Pull request by nandajavarma**
## Description

This PR adds a `werft` flag to create self-hosted preview environments from PRs. The idea is that, users will be able to make the following comment to the PR and a werft job will get triggered from the build job (with status written back to Github) creating a preview environment:
```
/werft run with-sh-preview
```
The above command will do the following:
1. Publish the branch to KOTS
2. Create a replicated license for the KOTS channel created in the previous step
3. Start a self-hosted preview on a newly created k3s cluster(we are planning to work on reducing waste here)
4. Delete the created license
5. The URL of the preview setup will be available under the Results tab of the werft job

If the user wants to test on cluster other than `k3s`, they can run:

```
werft run with-sh-preview cluster=aks # for Azure k8s cluster
werft run with-sh-preview cluster=eks # for AWS k8s cluster
werft run with-sh-preview cluster=gke # for GCP k8s cluster
```

The created cluster will be deleted in 10 hours.

## Related Issue(s)

Fixes #7824

## How to test

Currently, since this change is not merged with `main`, one cannot add comments to the PR to trigger the builds. Instead you will have to open a gitpod workspace from this PR and run:

```
werft run github -j .werft/build.yaml -a with-sh-preview=true -a cluster=k3s
# or
werft run github -j .werft/build.yaml -a with-sh-preview=true -a cluster=aks
# or
werft run github -j .werft/build.yaml -a with-sh-preview=true -a cluster=eks
# or
werft run github -j .werft/build.yaml -a with-sh-preview=true -a cluster=gke
```

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-preview

### CODE REVIEW

**Comment by roboquat** [comment]
@nandajavarma: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by nandajavarma** [comment]
Hey @meysholdt ! Yes! it is related to the RFC, yes. The goal of this PR is to do 2 things: 1) being able to start-up preview setups from PRs to improve the dev velocity especially for the self-hosted team and 2) workflow to create and delete replicated licence that can be used to try out newer changes. 

Once this PR is merged, folks should be able to comment `/werft run with-sh-preview` to create self-hosted preview setup on k3s, which will get cleaned up after 10 hours. Goal was to get it merged before the release testing starts. Basically a big part of #7824 will be addressed by this. Let me know if that sheds a bit more light  on the PR.

**Comment by nandajavarma** [comment]
@adrienthebo since this is not a part of main yet, the flags as comment doesn't work yet. You will have to follow the `How to test` section to trigger them

**Review by adrienthebo — COMMENTED** [review]
I believe I've found an issue - though it might be outside of the juristiction of this pull request. Error is as follows (from [gitpod-self-hosted-installer-tests-head.77](https://werft.gitpod-dev.com/job/gitpod-self-hosted-installer-tests-head.77))

```
Error: Activated service account credentials for: [sh-automated-tests@sh-automated-tests.iam.gserviceaccount.com]
Fetching cluster endpoint and auth data.
ERROR: (gcloud.container.clusters.get-credentials) ResponseError: code=404, message=Not found: projects/sh-automated-tests/zones/europe-west1-d/clusters/gp-nvn-sh-prev-k3s.
No cluster named 'gp-nvn-sh-prev-k3s' in sh-automated-tests.
Activated service account credentials for: [sh-automated-tests@sh-automated-tests.iam.gserviceaccount.com]
CommandException: No URLs matched: gs://nightly-tests/tf-state/nvn-sh-prev-k3s-kubeconfig
╷
│ Error: Error, failed to create instance sql-nvn-sh-prev-k3s: googleapi: Error 409: The Cloud SQL instance already exists. When you delete an instance, you can't reuse the name of the deleted instance until one week from the deletion date., instanceAlreadyExists
│
│   with module.k3s.google_sql_database_instance.gitpod,
│   on ../infra/modules/k3s/main.tf line 175, in resource "google_sql_database_instance" "gitpod":
│  175: resource "google_sql_database_instance" "gitpod" {
│
╵
make: *** [Makefile:129: k3s-standard-cluster] Error 1

Trying to send slack alert
tracing warning: No slice span by name create-gcp-infra
```

AKS is failing, again looks unrelated

```
Error: ╷
│ Error: creating Cluster: (Managed Cluster Name "gitpod-test-nor-primary-nvnshprevaks" / Resource Group "gitpod-test-nor-nvnshprevaks"): containerservice.ManagedClustersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="UnsupportedAliasMinorVersion" Message="1.21 is not a supported Kubernetes version. To see our supported AKS version list please visit https://docs.microsoft.com/azure/aks/supported-kubernetes-versions?tabs=azure-cli for more details."
│
│   with module.aks.azurerm_kubernetes_cluster.k8s,
│   on ../infra/modules/aks/kubernetes.tf line 17, in resource "azurerm_kubernetes_cluster" "k8s":
│   17: resource "azurerm_kubernetes_cluster" "k8s" {
│
╵
make: *** [Makefile:97: aks-standard-cluster] Error 1

Trying to send slack alert
tracing warning: No slice span by name create-azure-infra
```

**Review by adrienthebo — CHANGES_REQUESTED** [review]
As noted in an earlier comment, we have test failures on all platforms that prevent this from reliably working out of the gate. I've opened https://github.com/gitpod-io/gitpod/pull/12321/files targeting this branch with the fixups, let's give those fixes a try (or equivalent changes) and see if that improves the reliability. Looking forward to getting this merged!

File: install/infra/modules/gke/main.tf
Comment (nandajavarma) [reply install/infra/modules/gke/main.tf:158]: fixed by the other PR

**Comment by nandajavarma** [comment]
@adrienthebo All the dependent PRs has been reviewed and merged. Can we try and merge this one ?

File: install/tests/cleanup.sh
Comment (MrSimonEmms) [inline install/tests/cleanup.sh:39]: Is there any way of making this a function that can be invoked? This is a repetition of what's in `.github/workflows/delete-kots-channel.yml`

**Review by MrSimonEmms — APPROVED** [review]
/hold

Looks good. I have one question - if the answer is "no", then feel free to remove the hold and get this merged

File: install/tests/cleanup.sh
Comment (nandajavarma) [reply install/tests/cleanup.sh:39]: we ideally could, but I don't know if it is much of an optimization considering it means we will have to pass the API token, license name and in the very near future appname since I have an issue open to use gitpod-dev as published channel. Do you mind if I think about this better when I deal with that issue?

Comment (MrSimonEmms) [reply install/tests/cleanup.sh:39]: Absolutely - let's get it merged

**Pull request merged**
